java/code/src/com/suse/manager/clusters/ClusterManager.java

```diff
@@ -0,0 +1,609 @@
+     */
+    public static Optional<ClusterProvider> findClusterProvider(String label) {
+        return findClusterProviders().stream()
+                .filter(provider -> label.equals(provider.getLabel()))
+                .findFirst();
+    }
+
+
+    /**
+     * Queries the cluster to get the all the nodes and tries to match
+     * them with the registered systems.
+     * @param cluster the cluster
+     * @return a list of cluster nodes
+     */
+    public List<ClusterNode> listClusterNodes(Cluster cluster) {
+        List<ClusterNode> result = new ArrayList<>();
+
+        Optional<Map<String, Object>> settingsFormulaData = formulaManager
+                .getClusterFormulaData(cluster, "settings");
+        if (settingsFormulaData.isEmpty()) {
+            throw new RuntimeException("No settings data found for cluster " + cluster.getLabel());
+        }
+
+        ClusterProviderParameters cpp =
+                new ClusterProviderParameters(cluster.getProvider(), Optional.of(settingsFormulaData.get()));
+        systemQuery.listClusterNodes(cluster.getManagementNode(), cpp).ifPresent(ret -> {
+            ret.forEach((k, v) -> result.add(new ClusterNode(k, v)));
+        });
+        matchClusterNodes(result);
+        return result;
+    }
+
+    /**
+     * Get the nodes that can be joined to the given cluster. For each node it checks the preconditions for joining
+     * a cluster required by the cluster provider.
+     *
+     * @param cluster the cluster
+     * @param user the user
+     * @return a Map with minion as keys and error/warning messages as values
+     */
+    public Map<MinionServer, List<Tuple2<String, String>>> getNodesAvailableForJoining(Cluster cluster, User user) {
+        return MinionServerFactory.lookupVisibleToUser(user)
+                .filter(server -> !ClusterFactory.listAllManagementNodes().contains(server))
+                .filter(server -> !ClusterFactory.listAllClustersNodes().contains(server))
+                .collect(Collectors.toMap(Function.identity(), minion -> checkJoinPreconditions(cluster, minion)));
+        }
+
+    private List<Tuple2<String, String>> checkJoinPreconditions(Cluster cluster, MinionServer minion) {
+        Optional<List<String>> requiredPackages = FormulaFactory
+                .getClusterProviderMetadata(cluster.getProvider(), "channels:required_packages", List.class)
+                .map(m -> (List<String>)m);
+        var msgs = new ArrayList<Tuple2<String, String>>();
+        requiredPackages.ifPresent(pkgs ->
+            pkgs.forEach(pkgName -> {
+                boolean pkgInChannels =
+                        Optional.ofNullable(PackageManager.lookupEvrIdByPackageName(minion.getId(), pkgName))
+                        .map(res -> !res.isEmpty())
+                        .orElse(false);
+                if (!pkgInChannels) {
+                    msgs.add(new Tuple2<>("warning", "Package " + pkgName + " not found in subscribed channels"));
```

Comment:
We should use `LocalizationService` here

Reply:
Fixed